### PR TITLE
Add checkpointing support for (most) fold-grammars programs

### PR DESCRIPTION
**Pull request by fymue**
To enable fold-grammars programs such as e.g. `pKiss` to be [checkpointed](https://github.com/jlab/gapc/pull/163), I added `serialize` methods to most external types defined in [Extensions](https://github.com/jlab/fold-grammars/tree/master/Extensions) and also updated the `rnaoptions.hh` file so the user-compiled binaries get supplied with all checkpointing-related options.

I was unfortunately unable to support all of the fold-grammars programs, because some external types (e.g. `struct **_macrostate` types) contain `Subsequence` objects, for which type accessors need to be generated, which cannot automatically be generated for external types and also cannot be reliably done manually/by hand as there could be multiple accessors in an algebra product, which could break any accessors defined manually.
So for reliability reasons, I chose to not support certain external types and thus also to not support certain fold-grammars programs.

Most fold-grammars programs are still supported though.

### CODE REVIEW

**Comment by sjanssen2** [comment]
> I was unfortunately unable to support all of the fold-grammars programs, because some external types (e.g. `struct **_macrostate` types)

can you list all external datatypes that are affected, such that I can easily check?

File: Extensions/rnaoptions.hh [hunk 1]
```diff
@@ -137,6 +138,8 @@ class Opts {
     size_t checkpoint_interval;  // default interval: 3600s (1h)
     boost::filesystem::path  checkpoint_out_path;  // default path: cwd
     boost::filesystem::path  checkpoint_in_path;  // default: empty
+	std::string user_file_prefix;
```
Comment (sjanssen2) [inline Extensions/rnaoptions.hh:141]: is the extra indent intentional?

File: Extensions/rnaoptions.hh [hunk 2]
```diff
@@ -295,15 +300,14 @@ class Opts {
...
-                << "   command line inputs as the inputs to this binary to ensure the correctness of the answer."
-				<< std::endl << std::endl
+                << "-O, --checkpointOutput <path/prefix> set path where to store the checkpoints; default: current working directory" << std::endl
+                << "   Optional: add custom prefix for generated files to path" << std::endl
+				<< "   (e.g. path: \"path/to/dir/file_prefix\" will set path to \"/path/to/dir/\" and prefix to \"file_prefix\")." << std::endl
+				<< "   Make sure to add a \"/\" to the end of path if you don't wish to add a custom prefix to the files." << std::endl << std::endl
+                << "-I, --checkpointInput <logfile> set the path to the Logfile of the checkpoints you wish to load." << std::endl
+				<< "   (This file was generated along with the checkpoint archives." << std::endl
+				<< "    If it isn't available add the path to each archive to a text file and provide the path to this file)." << std::endl << std::endl
+				<< "-K, --keepArchives don't delete checkpointing archives after the program finished its calculations" << std::endl << std::endl
```
Comment (sjanssen2) [inline Extensions/rnaoptions.hh:310]: this is copy and paste from gapc right?

File: Extensions/typesRNAfolding.hh [hunk 3]
```diff
@@ -1,7 +1,22 @@
...
+#endif
+
 struct answer_pknot_mfe {
+#ifdef CHECKPOINTING_INTEGRATED
+    friend class boost::serialization::access;
+
+	template <class Archive>
+	void serialize(Archive &ar, const unsigned int version) {
+	    ar & energy;
+		ar & betaLeftOuter;
```
Comment (sjanssen2) [inline Extensions/typesRNAfolding.hh:15]: extra indent

**Review by sjanssen2 — CHANGES_REQUESTED** [review]
a test in https://github.com/jlab/fold-grammars/blob/1d3d8e007c0cb52e51dbe32681eae6c5393b3e7c/Misc/Test-Suite/GeorgStyle/config#L180 would be great

File: Extensions/rnaoptions.hh (hunk 1 above)
Comment (fymue) [reply Extensions/rnaoptions.hh:141]: > can you list all external datatypes that are affected, such that I can easily check? Maybe I can refactor such that only int indices are provided but not the actual sequence

`mfecovar_macrostate`, `answer_ali_pfunc_macrostate`, `answer_macrostate_mfe` and `answer_macrostate_pfunc` (defined in [typesRNAfolding.hh](https://github.com/jlab/fold-grammars/blob/master/Extensions/typesRNAfolding.hh))

**Comment by fymue** [comment]
> can you list all external datatypes that are affected, such that I can easily check? Maybe I can refactor such that only int indices are provided but not the actual sequence

`mfecovar_macrostate`, `answer_ali_pfunc_macrostate`, `answer_macrostate_mfe` and `answer_macrostate_pfunc` (defined in [typesRNAfolding.hh](https://github.com/jlab/fold-grammars/blob/master/Extensions/typesRNAfolding.hh))

File: Extensions/rnaoptions.hh (hunk 2 above)
Comment (fymue) [reply Extensions/rnaoptions.hh:310]: Yes.

File: Extensions/rnaoptions.hh (hunk 1 above)
Comment (fymue) [reply Extensions/rnaoptions.hh:141]: No it isn't. I will fix that.

**Comment by fymue** [comment]
> a test in
> 
> https://github.com/jlab/fold-grammars/blob/1d3d8e007c0cb52e51dbe32681eae6c5393b3e7c/Misc/Test-Suite/GeorgStyle/config#L180
> would be great

Alright, I will create 1 or 2 tests then.

**Comment by fymue** [comment]
I actually just realized that there are all kinds of indent problems in both files, because spaces and tabs were mixed in both. Should we maybe add those files to the `cpplint` check and have me replace all tabs with spaces @sjanssen2?

**Comment by fymue** [comment]
> 

I fixed all indent/linter errors for both files and pushed the changes to a new branch so the `diff` view doesn't explode on here and we can keep this issue separate from this PR. I suggest we just ignore all the indent errors that are currently present in the files on this branch for now. After this PR can be merged, I will simply create a new PR from the other branch where all linter errors are fixed.

**Review by sjanssen2 — APPROVED** [review]

**Comment by sjanssen2** [comment]
> I actually just realized that there are all kinds of indent problems in both files, because spaces and tabs were mixed in both. Should we maybe add those files to the `cpplint` check and have me replace all tabs with spaces @sjanssen2?

I would like that a lot. Thank you! We should maybe also extend the github workflow for automatic linting of these files?!
https://github.com/jlab/gapc/blob/caae6891d3f6afb2dbabd55adf84d2cd112b7546/.github/workflows/c-cpp.yml#L141-L149

**Comment by fymue** [comment]
> I would like that a lot. Thank you! We should maybe also extend the github workflow for automatic linting of these files?! https://github.com/jlab/gapc/blob/caae6891d3f6afb2dbabd55adf84d2cd112b7546/.github/workflows/c-cpp.yml#L141-L149

Already done. I will open the PR for that as soon as this one is merged.

**Comment by sjanssen2** [comment]
let me come back to 
> contain Subsequence objects, for which type accessors need to be generated, which cannot automatically be generated for external types
The datatype `Subsequence` itself should be defined internally withing gapc. It holds `i` and `j` as left and right border of an infix of the user input sequence and `seq` as the full string of the user input. I did not yet understand why this causes the issues. I could check if it is possible to replace with a more simple structure only holding `i` and `j`. Would that help?

**Comment by fymue** [comment]
> The datatype `Subsequence` itself should be defined internally withing gapc. It holds `i` and `j` as left and right border of an infix of the user input sequence and `seq` as the full string of the user input. I did not yet understand why this causes the issues. I could check if it is possible to replace with a more simple structure only holding `i` and `j`. Would that help?

The issue/problem with this is that it would be extremely inefficient to serialize the data that `seq` points to, because, as you've mentioned, it is shared across all `Subsequence` objects and points to the whole input sequence, which would lead to really big archives and slow dump times if I also archive this data.

What I do to avoid this is I just serialize the `i` and `j` values,  loop over all `Subsequence` objects after loading the archives and simply set the (at this point still empty) `seq` pointer to the input sequence of the program (which has to be the same as the one that the archives were created with, otherwise an error would've been thrown already). I figured that this is the most efficient way of solving this.

However, this obviously means that I need to know how to access all `Subsequence` objects in the loop, for which I need to generated accessors, which I can't generate for external types. I hope this explains the problem.

So yes, it would be really helpful if you could find a way to switch out these types for a type that only contains the `i` and `j` positions, as this type would be easily serializable.

**Review by kmaibach — APPROVED** [review]

**Pull request merged**